test/functional/normal/put_spec.lua

```diff
@@ -0,0 +1,1012 @@
+                )
+            end)
+        end)
+
+
+        describe("with 'virtualedit'", function()
+            describe('splitting a tab character', function()
+                local base_expect_string = [[
+                Line of words 1
+                  test_stringx"     Line of words 2]]
+                run_normal_mode_tests(
+                    base_expect_string,
+                    'vp',
+                    function()
+                        funcs.setline('$', '	Line of words 2')
+                        execute('set virtualedit=all')
+                        funcs.setpos('.', {0, 2, 1, 2, 3})
+                    end,
+                    nil,
+                    ' '
+                )
+            end)
+            describe('after end of line', function()
+                local base_expect_string = [[
+                Line of words 1  test_stringx"
+                Line of words 2]]
+                run_normal_mode_tests(
+                    base_expect_string,
+                    'vp',
+                    function()
+                        execute('set virtualedit=all')
+                        funcs.setpos('.', {0, 1, 16, 2, 18})
+                    end,
+                    true,
+                    ' '
+                )
+            end)
+        end)
+    end)
+
+    describe('. register special tests', function()
+        before_each(reset)
+        it('applies control character actions', function()
+            feed('i<C-t><esc>u')
+            expect([[
+            Line of words 1
+            Line of words 2]])
+            feed('".p')
+            expect([[
+            	Line of words 1
+            Line of words 2]])
+            feed('u1go<C-v>j".p')
+            eq([[
+	ine of words 1
+	ine of words 2]], curbuf_contents())
+        end)
+
+        it('should not ring the bell with gp at end of line', function()
+            local Screen = require('test.functional.ui.screen')
+            local screen = Screen.new()
+            screen:attach()
+            helpers.ok(not screen.bell and not screen.visualbell)
+            execute('set beloff=""')
+            execute('goto 1')
+            feed('$".gp')
+            screen:wait(function()
+                if screen.bell or screen.visualbell then
+                    return 'Bell was rung after paste'
+                end
+            end)
+            helpers.ok(not screen.bell and not screen.visualbell)
+            screen:detach()
+        end)
+        it('should restore cursor position after undo of ".p', function()
+            local origpos = funcs.getcurpos()
+            feed('".pu')
+            eq(origpos, funcs.getcurpos())
+        end)
+    end)
+
+    describe('register . problems', function()
+        before_each(reset)
+        -- I don't think these are reasonable to fix -- usually because the fix
+        -- adds too much interdependence for what it's worth.
+        -- I'm leaving them here for documentation.
```

Comment:
For the cases we don't plan to fix, I suggest pasting this entire `describe('register . problems' ...` source into the commit message.